RFC-0011-coordinator-throttling.md

```diff
@@ -0,0 +1,172 @@
+
+While not exhaustive, current throttling/possible mechanisms relevant for this discussion
+
+#### Granular task based scheduling
+
+**Leaf Task scheduling**
+
+The node selection logic in Presto involves randomly selecting 10 nodes from those that do not currently have scheduled splits for the given query. Among these 10 nodes, the system then chooses the node that is the least busy. For reference, the relevant code implementation can be found [here](https://github.com/prestodb/presto/blob/master/presto-main-base/src/main/java/com/facebook/presto/execution/scheduler/nodeSelection/SimpleNodeSelector.java#L182).
+Additionally, in the process of leaf task scheduling, a feedback mechanism is employed whereby splits are assigned to the nodes deemed least busy. It is important to note that in this context, "busyness" is defined by the number of splits being processed by a node, rather than its actual system load. 
+
+While this approach is generally effective for distributing work, it may not always result in the most optimal allocation of resources.
+
+**Intermediate task scheduling**
+
+For intermediate task scheduling in Presto, the selection of workers is based on a hash partition count. Specifically, intermediate tasks are assigned to worker nodes according to the hash of the partition, which determines the distribution across available intermediate workers. The relevant code implementation can be found [here](https://github.com/prestodb/presto/blob/master/presto-main-base/src/main/java/com/facebook/presto/sql/planner/SystemPartitioningHandle.java#L164).
+This strategy does not consider the current system load when assigning intermediate tasks. 
+
+#### Resource Group throttling
+There are certain metrics like softMemoryLimit etc which exist at RG level to queue the workload, but it does not consider the actual utilization and is out of scope for this discussion.
+
+
+While we could improve task scheduling (pro-active measure to not let worker overload) through load balancing (more on this in alternative approaches), this alone would be insufficient given the nature of our overload scenario where all workers are saturated in short period of time which is due to admitting lot of heavy queries in short duration of time. 
+
+For immediate relief, we propose implementing a holistic admission control mechanism that provides reactive throttling of the overall workload by not admitting queries.
+
+### Goals
+- Decrease duration of overload on the cluster
+- Also decrease overall variance of execution time of queries (queuing is fine)
+
+## Proposed Implementation
+### Worker side changes
+- Added endpoint to gather worker load (`/v1/info/nodestats`)
+  - Include NodeState + worker load. Idea is to just call `/v1/info/nodestats` rather than `/v1/info/state` (eventually deprecate it). Worker load is already getting populated in [cpp workers](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/PrestoServer.cpp#L1537).
```

Comment:
Can this plan include the deprecation?  I think we need to clean up the old endpoint as well so we're not confusingly returning two different things depending on whether or not we're running a C++ cluster.

Reply:
Added the deprecation plan